query-engine/core/src/query_graph_builder/write/upsert.rs

```diff
@@ -177,3 +191,86 @@ pub fn upsert_record(
 
     Ok(())
 }
+
+// This optimisation on our upserts allows us to use the `INSERT ... ON CONFLICT SET ..`
+// when the query matches the following conditions:
+// 1. The data connector supports it
+// 2. The create and update arguments do not have any nested queries
+// 3. There is only 1 unique field in the where clause
+// 4. The unique field defined in where clause has the same value as defined in the create arguments
+fn can_use_connector_native_upsert(
+    model: &ModelRef,
+    where_field: ParsedInputMap,
+    create_argument: ParsedInputMap,
+    update_argument: ParsedInputMap,
+    connector_ctx: &ConnectorContext,
+) -> bool {
+    let has_nested_create = create_argument
+        .clone()
+        .into_iter()
+        .any(|(k, _)| matches!(model.fields().find_from_all(&k), Ok(Field::Relation(_))));
+
+    let has_nested_update = update_argument
+        .into_iter()
+        .any(|(k, _)| matches!(model.fields().find_from_all(&k), Ok(Field::Relation(_))));
+
+    let uniques = where_field
+        .clone()
+        .into_iter()
+        .filter(|(field_name, input)| {
+            is_unique_field(field_name, model) && where_and_create_equal(field_name, input, &create_argument)
+        })
+        .count();
+
+    let where_values_same_as_create = where_field
+        .clone()
+        .into_iter()
+        .all(|(field_name, input)| where_and_create_equal(&field_name, &input, &create_argument));
+
+    connector_ctx.can_native_upsert()
+        && uniques == 1
+        && !has_nested_create
+        && !has_nested_update
+        && where_values_same_as_create
+}
+
+fn is_unique_field(field_name: &String, model: &ModelRef) -> bool {
+    match model.fields().find_from_scalar(&field_name) {
+        Ok(field) => field.unique(),
+        Err(_) => model
+            .unique_indexes()
+            .iter()
+            .any(|index| index.name == Some(field_name.to_string())),
+    }
+}
```

Comment:
I think this should use the `resolve_compound_field` function instead, located here https://github.com/prisma/prisma-engines/blob/1a819bfa00df930fab0fb30f0eb52a6d731ee128/query-engine/core/src/query_graph_builder/extractors/utils.rs#L5-L7.
First to dry things more but more importantly because this logic only looks for `@@unique` compound but not for `@@id` compounds.